packages/utils/telemetry-utils/src/sampledTelemetryHelper.ts

```diff
@@ -0,0 +1,114 @@
+     * @param logger - The logger to use to write the telemetry performance event.
+     * @param sampleThreshold - Telemetry performance events will be generated every time we hit this many
+     *                          executions of the code block.
+     * @param includeAggregateMetrics - If set to `true`, the telemetry performance event will include aggregated
+     *                                  metrics (execution count, total duration, min duration, max duration) for
+     *                                  all the executions in between generated events.
+     */
+    public constructor(
+        private readonly eventBase: ITelemetryGenericEvent,
+        private readonly logger: ITelemetryLogger,
+        private readonly sampleThreshold: number,
+        private readonly includeAggregateMetrics: boolean = false) {
+    }
+
+    /**
+     * @param codeToMeasure - The code to be executed and measured.
+     * @param dimension - A key to track executions of the code block separately. Each different
+     *                    value of this parameter has a separate set of executions and metrics tracked
+     *                    by the class. If no such distinction needs to be made, do not provide a value.
+     * @returns Whatever the passed-in code block returns.
+     */
+    public measure<T>(codeToMeasure: () => T, dimension: string = ""): T {
+        const start = performance.now();
+        const returnValue = codeToMeasure();
+        const duration = performance.now() - start;
+
+        let m = this.measurementsMap.get(dimension);
+        if (m === undefined) {
+            m = { count: 0, measurements: { duration: 0 } };
+            this.measurementsMap.set(dimension, m);
+        }
+        m.count++;
+        m.measurements.duration = duration;
+
+        if (this.includeAggregateMetrics) {
+            m.measurements.count = (m.measurements.count ?? 0) + 1;
+            m.measurements.totalDuration = (m.measurements.totalDuration ?? 0) + duration;
+            m.measurements.minDuration = Math.min(m.measurements.minDuration ?? Infinity, duration);
+            m.measurements.maxDuration = Math.max(m.measurements.maxDuration ?? 0, duration);
+        }
+
+        if (m.count >= this.sampleThreshold) {
+            this.flushDimension(dimension);
+        }
+
+        return returnValue;
+    }
+
+    private flushDimension(dimension: string) {
+        const measurements = this.measurementsMap.get(dimension);
+        if (measurements === undefined) {
+            return;
+        }
+
+        if (measurements.count !== 0) {
+            const telemetryEvent: ITelemetryPerformanceEvent = {
+                ...this.eventBase,
+                ...measurements.measurements,
+                dimension,
+            };
+
+            this.logger.sendPerformanceEvent(telemetryEvent);
+            this.measurementsMap.set(dimension, {
```

Comment:
nit: since this called on dispose its a bit odd to re-populate the map, especially as we do have delay initialization, so i think it would be fine to just delete the key